evalml/pipelines/components/transformers/encoders/ordinal_encoder.py

```diff
@@ -0,0 +1,320 @@
+            for col in X_t[self.features_to_encode]:
+                ltype = ww_logical_types[col]
+                # Copy the order list, since we might mutate it later by adding nans
+                # and don't want to impact the Woodwork types
+                column_ordered_categories = ltype.order.copy()
+
+                if top_n is None or len(column_ordered_categories) <= top_n:
+                    unique_values = column_ordered_categories
+                else:
+                    # Don't count nans as a value
+                    value_counts = X_t[col].value_counts(dropna=True).to_frame()
+                    value_counts = value_counts.sample(
+                        frac=1,
+                        random_state=self._initial_state,
+                    )
+                    value_counts = value_counts.sort_values(
+                        [col],
+                        ascending=False,
+                        kind="mergesort",
+                    )
+                    unique_values = value_counts.head(top_n).index.tolist()
+                    # Categories should be in the same order as the data's Ordinal.order categories
+                    # instead of sorted by most frequent
+                    unique_values = [
+                        cat for cat in column_ordered_categories if cat in unique_values
+                    ]
+
+                categories.append(unique_values)
+
+        # Add any null values into the categories lists so that they aren't treated as unknown values
+        # This is needed because Ordinal.order won't indicate if nulls are present, and SKOrdinalEncoder
+        # requires any null values be present in the categories list if they are to be encoded as
+        # missing values
+        if isinstance(categories, list):
+            for i, col in enumerate(X_t[self.features_to_encode]):
+                if X_t[col].isna().any():
+                    categories[i] += [np.nan]
+
+        encoded_missing_value = self.parameters["encoded_missing_value"]
+        if encoded_missing_value is None:
+            encoded_missing_value = np.nan
+
+        self._encoder = SKOrdinalEncoder(
+            categories=categories,
+            handle_unknown=self.parameters["handle_unknown"],
+            unknown_value=self.parameters["unknown_value"],
+            encoded_missing_value=encoded_missing_value,
+        )
+
+        self._encoder.fit(X_t[self.features_to_encode])
+        return self
+
+    def transform(self, X, y=None):
+        """Ordinally encode the input data.
+
+        Args:
+            X (pd.DataFrame): Features to encode.
+            y (pd.Series): Ignored.
+
+        Returns:
+            pd.DataFrame: Transformed data, where each ordinal feature has been encoded into
+            a numerical column using where of ordinal integers represent
+            the relative order of categories.
+        """
+        X = infer_feature_types(X)
+
+        X_copy = X.ww.copy()
```

Comment:
Nitpick about variable names: I would replace the X with X_t and X_copy with X. We should also avoid the need to copy in that case, since calling `X.ww.drop` should return a new DataFrame.
```
X_t = X.ww.drop(columns=self.features_to_encode)
```

Reply:
Implemented this in dd5c075ff - happy to change if we should be doing this differently!

Reply:
I'm not sure what you mean here by "we don't want users messing around with the same object they passed in", could you explain that a little bit? If we're trying not to modify the user's original data, this wouldn't be the place to enforce it. I'm also worried that copying the data (especially when this component should be a no-op) uses unnecessary time and storage, especially with larger datasets.

Reply:
Yeah, I was assuming that this was the place to be worrying about modifying the user's original data. My thought was that if you pass a dataframe into `transform`, we shouldn't ever give you back the exact same object. But if that's not a contract we need to uphold here, I'm more than happy to avoid the extra computation!